File: substrate/settings_add_model_inventory_compose.py
"""Settings add-model inventory pack (pure).

secrets_stored, live_router_authorized, inventory_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from substrate.settings_decision_tree_usage_bar_compose import (
    SettingsDecisionTreeUsageBarCompose,
    SettingsDecisionTreeUsageBarComposeError,
    compose_settings_decision_tree_usage_bar,
)
from substrate.settings_model_inventory_budget_compose import (
    SettingsModelInventoryBudgetCompose,
    SettingsModelInventoryBudgetComposeError,
    compose_settings_model_inventory_budget,
)

AddModelAction = Literal["preview", "propose_add"]
VALID_ACTIONS = frozenset(("preview", "propose_add"))


class SettingsAddModelInventoryComposeError(ValueError):
    """Fail-closed validation for settings add-model inventory."""


@dataclass(frozen=True)
class SettingsAddModelInventoryCompose:
    inventory: SettingsModelInventoryBudgetCompose
    decision_tree: SettingsDecisionTreeUsageBarCompose | None
    action: AddModelAction
    proposed_new_model_ids: tuple[str, ...]
    proposed_new_count: int
    pack_ready: bool
    secrets_stored: bool
    live_router_authorized: bool
    inventory_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "inventory": self.inventory.to_dict(),
            "decision_tree": (
                self.decision_tree.to_dict() if self.decision_tree else None
            ),
            "action": self.action,
            "proposed_new_model_ids": list(self.proposed_new_model_ids),
            "proposed_new_count": self.proposed_new_count,
            "pack_ready": self.pack_ready,
            "secrets_stored": False,
            "live_router_authorized": False,
            "inventory_mutated": False,
            "notes": list(self.notes),
            "authority": "settings_add_model_inventory_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise SettingsAddModelInventoryComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _require_model_id(value: object, *, field: str) -> str:
    mid = _require_nonempty(value, field=field)
    import re

    if (
        len(mid) > 128
        or re.search(r"sk-|api[_-]?key|secret|bearer\s", mid, re.I)
        or " " in mid
    ):
        raise SettingsAddModelInventoryComposeError(
            f"{field} must be a model id, not secret material"
        )
    return mid


def compose_settings_add_model_inventory(
    *,
    models: object,
    pending_add_model_ids: object,
    action: object,
    daily_cap_usd: object,
    spent_usd: object,
    operator_ack: object,
    selected_model_id: object | None = None,
    projected_cost_usd_high: object | None = None,
    projected_cost_usd_low: object | None = None,
) -> SettingsAddModelInventoryCompose:
    """Add-model inventory propose/preview. Never stores secrets or mutates."""
    if not isinstance(operator_ack, bool):
        raise SettingsAddModelInventoryComposeError(
            "operator_ack must be an explicit boolean"
        )
    if action not in VALID_ACTIONS:
        raise SettingsAddModelInventoryComposeError(
            "action must be preview or propose_add"
        )
    action_s: AddModelAction = action  # type: ignore[assignment]
    if not isinstance(pending_add_model_ids, list):
        raise SettingsAddModelInventoryComposeError(
            "pending_add_model_ids must be an array"
        )

    notes: list[str] = [
        "secrets_stored=false — model ids only; never raw API keys",
        "live_router_authorized=false — operator selects model",
        "inventory_mutated=false — propose_add is intent only",
    ]

    try:
        inventory = compose_settings_model_inventory_budget(
            models=models,
            daily_cap_usd=daily_cap_usd,
            spent_usd=spent_usd,
            pending_add_model_ids=pending_add_model_ids,
            selected_model_id=selected_model_id,
        )
    except SettingsModelInventoryBudgetComposeError as e:
        raise SettingsAddModelInventoryComposeError(str(e)) from e
    notes.extend(f"[inventory] {n}" for n in inventory.notes)

    existing = set(inventory.model_ids)
    proposed_new: list[str] = []
    seen: set[str] = set()
    for i, raw in enumerate(pending_add_model_ids):
        mid = _require_model_id(raw, field=f"pending_add_model_ids[{i}]")
        if mid in seen:
            continue
        seen.add(mid)
        if mid not in existing:
            proposed_new.append(mid)
    notes.append(
        f"proposed_new_count={len(proposed_new)} (not yet in inventory)"
    )

    decision_tree: SettingsDecisionTreeUsageBarCompose | None = None
    if not isinstance(models, list):
        raise SettingsAddModelInventoryComposeError("models must be an array")
    model_options: list[dict[str, Any]] = []
    for m in models:
        if isinstance(m, dict) and m.get("model_id"):
            model_options.append(
                {
                    "model_id": m.get("model_id"),
                    "tier": m.get("tier"),
                }
            )
    if action_s == "propose_add":
        for mid in proposed_new:
            model_options.append({"model_id": mid, "tier": "pending_add"})

    if model_options:
        if selected_model_id is not None and str(selected_model_id).strip():
            selected = _require_model_id(
                selected_model_id, field="selected_model_id"
            )
        else:
            selected = str(model_options[0]["model_id"])
        ids = {str(m["model_id"]) for m in model_options}
        selected_final = selected if selected in ids else str(
            model_options[0]["model_id"]
        )
        try:
            decision_tree = compose_settings_decision_tree_usage_bar(
                selected_model_id=selected_final,
                models=model_options,
                daily_cap_usd=daily_cap_usd,
                spent_usd=spent_usd,
                operator_ack=operator_ack,
                projected_cost_usd_high=projected_cost_usd_high,
                projected_cost_usd_low=projected_cost_usd_low,
                pending_add_model_ids=pending_add_model_ids,
            )
        except SettingsDecisionTreeUsageBarComposeError as e:
            raise SettingsAddModelInventoryComposeError(str(e)) from e
        notes.extend(f"[decision] {n}" for n in decision_tree.notes)
    else:
        notes.append(
            "decision_tree skipped — empty inventory and no pending adds"
        )

    if action_s == "preview":
        pack_ready = operator_ack is True and inventory.secrets_stored is False
        notes.append(
            "pack_ready=true — inventory preview advisory"
            if pack_ready
            else "pack_ready=false — operator_ack required for preview pack"
        )
    else:
        if len(proposed_new) == 0:
            notes.append(
                "pack_ready=false — propose_add requires ≥1 new model id "
                "not already inventoried"
            )
            pack_ready = False
        elif not operator_ack:
            notes.append(
                "pack_ready=false — propose_add requires operator_ack"
            )
            pack_ready = False
        else:
            pack_ready = True
            notes.append(
                "pack_ready=true — propose_add intent ready; "
                "inventory_mutated=false"
            )

    if inventory.secrets_stored is not False or inventory.live_router_authorized is not False:
        raise SettingsAddModelInventoryComposeError(
            "invariant: inventory honesty flags must remain false"
        )
    if decision_tree is not None and (
        decision_tree.secrets_stored is not False
        or decision_tree.live_router_authorized is not False
    ):
        raise SettingsAddModelInventoryComposeError(
            "invariant: decision_tree honesty flags must remain false"
        )

    notes.extend(
        (
            "secrets_stored=false",
            "live_router_authorized=false",
            "inventory_mutated=false",
        )
    )

    return SettingsAddModelInventoryCompose(
        inventory=inventory,
        decision_tree=decision_tree,
        action=action_s,
        proposed_new_model_ids=tuple(proposed_new),
        proposed_new_count=len(proposed_new),
        pack_ready=pack_ready,
        secrets_stored=False,
        live_router_authorized=False,
        inventory_mutated=False,
        notes=tuple(notes),
        authority="settings_add_model_inventory_compose_advisory",
    )


def format_settings_add_model_inventory_summary(
    c: SettingsAddModelInventoryCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · action={c.action} · "
        f"inventory={c.inventory.inventory_count} · "
        f"proposed_new={c.proposed_new_count} · "
        f"secrets_stored=false · inventory_mutated=false · "
        f"live_router_authorized=false"
    )


__all__ = [
    "SettingsAddModelInventoryCompose",
    "SettingsAddModelInventoryComposeError",
    "compose_settings_add_model_inventory",
    "format_settings_add_model_inventory_summary",
]

File: apps/reading/src/api/settingsAddModelInventoryCompose.ts
/**
 * Settings add-model inventory pack (pure).
 *
 * Operator vision: add models in settings (BYOK inventory ids only), see
 * budget usage bar, and optionally select a driver — never store secrets,
 * never auto-route.
 *
 * secrets_stored always false.
 * live_router_authorized always false.
 * inventory_mutated always false (pure propose only).
 */

import {
  composeSettingsModelInventoryBudget,
  type InventoryModelRow,
  type SettingsModelInventoryBudgetCompose,
} from "./settingsModelInventoryBudgetCompose";
import {
  composeSettingsDecisionTreeUsageBar,
  type SettingsDecisionTreeUsageBarCompose,
} from "./settingsDecisionTreeUsageBarCompose";
import type { ModelOption } from "./modelDecisionPromptCompose";

export type AddModelAction = "preview" | "propose_add";

export interface SettingsAddModelInventoryInput {
  /** Current inventory (ids only). */
  models: InventoryModelRow[];
  /**
   * Model ids operator wants to add (BYOK). Rejected if secret-like.
   * For propose_add, these become the proposed inventory extension.
   */
  pending_add_model_ids: string[];
  action: AddModelAction;
  daily_cap_usd: number | null;
  spent_usd: number | null;
  selected_model_id?: string | null;
  projected_cost_usd_high?: number | null;
  projected_cost_usd_low?: number | null;
  operator_ack: boolean;
}

export interface SettingsAddModelInventoryCompose {
  inventory: SettingsModelInventoryBudgetCompose;
  decision_tree: SettingsDecisionTreeUsageBarCompose | null;
  action: AddModelAction;
  /** Ids that would be added (not already in inventory). */
  proposed_new_model_ids: string[];
  proposed_new_count: number;
  /**
   * True when operator_ack and (preview with valid inventory OR propose_add
   * with ≥1 new id). Still never mutates stored inventory.
   */
  pack_ready: boolean;
  secrets_stored: false;
  live_router_authorized: false;
  /** Always false — pure layer never writes settings inventory. */
  inventory_mutated: false;
  notes: string[];
  authority: "settings_add_model_inventory_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function requireModelId(value: unknown, name: string): string {
  const id = requireNonEmpty(value, name);
  if (
    id.length > 128 ||
    /sk-|api[_-]?key|secret|bearer\s/i.test(id) ||
    id.includes(" ")
  ) {
    throw new Error(`${name} must be a model id, not secret material`);
  }
  return id;
}

/**
 * Compose settings add-model inventory + optional decision tree pack.
 * Never stores secrets; never mutates inventory; never live-routes.
 */
export function composeSettingsAddModelInventory(
  input: SettingsAddModelInventoryInput,
): SettingsAddModelInventoryCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (input.action !== "preview" && input.action !== "propose_add") {
    throw new Error("action must be preview or propose_add");
  }
  if (!Array.isArray(input.pending_add_model_ids)) {
    throw new Error("pending_add_model_ids must be an array");
  }

  const notes: string[] = [
    "secrets_stored=false — model ids only; never raw API keys",
    "live_router_authorized=false — operator selects model",
    "inventory_mutated=false — propose_add is intent only",
  ];

  const inventory = composeSettingsModelInventoryBudget({
    models: input.models,
    pending_add_model_ids: input.pending_add_model_ids,
    daily_cap_usd: input.daily_cap_usd,
    spent_usd: input.spent_usd,
    selected_model_id: input.selected_model_id,
  });
  notes.push(...inventory.notes.map((n) => `[inventory] ${n}`));

  const existing = new Set(inventory.model_ids);
  const proposed_new_model_ids: string[] = [];
  const seen = new Set<string>();
  for (let i = 0; i < input.pending_add_model_ids.length; i++) {
    const id = requireModelId(
      input.pending_add_model_ids[i],
      `pending_add_model_ids[${i}]`,
    );
    if (seen.has(id)) continue;
    seen.add(id);
    if (!existing.has(id)) {
      proposed_new_model_ids.push(id);
    }
  }
  notes.push(
    `proposed_new_count=${proposed_new_model_ids.length} (not yet in inventory)`,
  );

  // Decision tree uses current inventory models; if propose_add and new ids,
  // surface them as ModelOptions for selection preview only.
  let decision_tree: SettingsDecisionTreeUsageBarCompose | null = null;
  const modelOptions: ModelOption[] = input.models.map((m) => ({
    model_id: m.model_id,
    tier: m.tier ?? undefined,
  }));
  if (input.action === "propose_add") {
    for (const id of proposed_new_model_ids) {
      modelOptions.push({ model_id: id, tier: "pending_add" });
    }
  }

  if (modelOptions.length > 0) {
    const selected =
      input.selected_model_id != null &&
      String(input.selected_model_id).trim() !== ""
        ? requireModelId(input.selected_model_id, "selected_model_id")
        : modelOptions[0].model_id;
    // selected must be in options for decision tree
    const ids = new Set(modelOptions.map((m) => m.model_id));
    const selected_final = ids.has(selected)
      ? selected
      : modelOptions[0].model_id;

    decision_tree = composeSettingsDecisionTreeUsageBar({
      selected_model_id: selected_final,
      models: modelOptions,
      daily_cap_usd: input.daily_cap_usd,
      spent_usd: input.spent_usd,
      projected_cost_usd_high: input.projected_cost_usd_high,
      projected_cost_usd_low: input.projected_cost_usd_low,
      pending_add_model_ids: input.pending_add_model_ids,
      operator_ack: input.operator_ack,
    });
    notes.push(...decision_tree.notes.map((n) => `[decision] ${n}`));
  } else {
    notes.push("decision_tree skipped — empty inventory and no pending adds");
  }

  let pack_ready = false;
  if (input.action === "preview") {
    pack_ready =
      input.operator_ack === true &&
      inventory.inventory_count >= 0 &&
      inventory.secrets_stored === false;
    notes.push(
      pack_ready
        ? "pack_ready=true — inventory preview advisory"
        : "pack_ready=false — operator_ack required for preview pack",
    );
  } else {
    // propose_add
    if (proposed_new_model_ids.length === 0) {
      notes.push(
        "pack_ready=false — propose_add requires ≥1 new model id not already inventoried",
      );
      pack_ready = false;
    } else if (!input.operator_ack) {
      notes.push("pack_ready=false — propose_add requires operator_ack");
      pack_ready = false;
    } else {
      pack_ready = true;
      notes.push(
        "pack_ready=true — propose_add intent ready; inventory_mutated=false",
      );
    }
  }

  if (
    inventory.secrets_stored !== false ||
    inventory.live_router_authorized !== false ||
    (decision_tree != null &&
      (decision_tree.secrets_stored !== false ||
        decision_tree.live_router_authorized !== false))
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("secrets_stored=false");
  notes.push("live_router_authorized=false");
  notes.push("inventory_mutated=false");

  return {
    inventory,
    decision_tree,
    action: input.action,
    proposed_new_model_ids,
    proposed_new_count: proposed_new_model_ids.length,
    pack_ready,
    secrets_stored: false,
    live_router_authorized: false,
    inventory_mutated: false,
    notes,
    authority: "settings_add_model_inventory_compose_advisory",
  };
}

export function formatSettingsAddModelInventorySummary(
  c: SettingsAddModelInventoryCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · action=${c.action} · ` +
    `inventory=${c.inventory.inventory_count} · ` +
    `proposed_new=${c.proposed_new_count} · ` +
    `secrets_stored=false · inventory_mutated=false · live_router_authorized=false`
  );
}
